File: src/main/java/com/HOIIVUtils/hoi4utils/clausewitz_data/country/Country.java
package com.HOIIVUtils.hoi4utils.clausewitz_data.country;

import com.HOIIVUtils.hoi4utils.clausewitz_map.state.InfrastructureData;
import com.HOIIVUtils.hoi4utils.clausewitz_map.state.State;
import com.HOIIVUtils.hoi4utils.clausewitz_map.buildings.Infrastructure;
import com.HOIIVUtils.hoi4utils.clausewitz_map.resources.Resources;
import com.HOIIVUtils.hoi4utils.clausewitz_data.technology.Technology;
import com.HOIIVUtils.hoi4utils.clausewitz_data.units.OrdersOfBattle;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

// todo make country extend countrytag???? ehhhhh
// todo consider... implements infrastructure, resources?????
// todo localizable data?
public class Country implements InfrastructureData, Comparable<Country> {
	private static final ObservableList<Country> countryList = FXCollections.observableArrayList();
	private CountryTag countryTag;
	private Infrastructure infrastructure;          // infrastructure of all owned states
	private Resources resources;                    // resources of all owned states
	private Set<OrdersOfBattle> oob_list;           // set of potential orders of battles defined in history/countries file (oob)
	private int defaultResearchSlots;               // default research slots as defined in history/countries file or similar

	private Set<CountryFlag> countryFlags;
	private int capital;                            // country capital as defined by applicable oob or unknown
	private double stability;                       // stability percentage defined from 0.0-1.0
	private double warSupport;                      // war support percentage defined from 0.0-1.0
	private Set<Technology> startingTech;           // starting technology defined in history/countries file


	public Country() {
		this(CountryTag.NULL_TAG);
	}

	public Country(CountryTag countryTag) {
		this(countryTag, new Infrastructure(), new Resources());
	}

	public Country(CountryTag countrytag, Infrastructure infrastructure, Resources resources) {
		this.countryTag = countrytag;
		this.infrastructure = infrastructure;
		this.resources = resources;
	}

	// todo if country extends country tag this could be much better
	public <T> Country(T item) {
		if (item instanceof CountryTag) {
			this.countryTag = (CountryTag) item;
			this.infrastructure = new Infrastructure();
			this.resources = new Resources();

		} else if (item instanceof Country country) {
			this.countryTag = country.countryTag;
			this.infrastructure = country.infrastructure;
			this.resources = country.resources;
		} else {
			this.countryTag = CountryTag.NULL_TAG;
			this.infrastructure = new Infrastructure();
			this.resources = new Resources();
		}
	}

	/**
	 * @return
	 * @deprecated
	 */
	@Deprecated
	public static List<Country> getList() {
		// TODO: Remove this deprecated code someday.
		if (countryList == null || countryList.isEmpty()) {
			loadCountries();
		}
		return countryList;
	}

	public static List<Function<Country, ?>> getCountryDataFunctions(boolean resourcePercentages) {
		List<Function<Country, ?>> dataFunctions = new ArrayList<>(18);         // 18 for optimization, limited number of data functions.

		dataFunctions.add(Country::name);
		dataFunctions.add(c -> c.infrastructure.population());
		dataFunctions.add(c -> c.infrastructure.civMilRatio());
		dataFunctions.add(c -> c.infrastructure.militaryFactories());
		dataFunctions.add(c -> c.infrastructure.navalDockyards());
		dataFunctions.add(c -> c.infrastructure.airfields());
		dataFunctions.add(c -> c.infrastructure.civMilRatio());
		dataFunctions.add(c -> c.infrastructure.popPerFactoryRatio());
		dataFunctions.add(c -> c.infrastructure.popPerCivRatio());
		dataFunctions.add(c -> c.infrastructure.popPerMilRatio());
		dataFunctions.add(c -> c.infrastructure.popAirportCapacityRatio());
		dataFunctions.add(c -> c.infrastructure.popPerStateRatio(c.numOwnedStates()));
		if (resourcePercentages) {
			dataFunctions.add(Country::aluminum);
			dataFunctions.add(Country::chromium);
			dataFunctions.add(Country::oil);
			dataFunctions.add(Country::rubber);
			dataFunctions.add(Country::steel);
			dataFunctions.add(Country::tungsten);
		} else {
			dataFunctions.add(Country::aluminumPercentOfGlobal);
			dataFunctions.add(Country::chromiumPercentOfGlobal);
			dataFunctions.add(Country::oilPercentOfGlobal);
			dataFunctions.add(Country::rubberPercentOfGlobal);
			dataFunctions.add(Country::steelPercentOfGlobal);
			dataFunctions.add(Country::tungstenPercentOfGlobal);
		}

		return dataFunctions;
	}

	public CountryTag countryTag() {
		return countryTag;
	}

	public void setCountryTag(CountryTag countryTag) {
		this.countryTag = countryTag;
	}

	@Override
	public Infrastructure getInfrastructureRecord() {
		return infrastructure;
	}

//	@Override
//	public int infrastructure() {
//		return infrastructure.infrastructure();
//	}
//
//	@Override
//	public int civilianFactories() {
//		return infrastructure.civilianFactories();
//	}
//
//	@Override
//	public int militaryFactories() {
//		return infrastructure.militaryFactories();
//	}
//
//	@Override
//	public int navalDockyards() {
//		return infrastructure.navalDockyards();
//	}
//
//	@Override
//	public int navalPorts() {
//		return infrastructure.navalPorts();
//	}
//
//	@Override
//	public int airfields() {
//		return infrastructure.airfields();
//	}

	public void setInfrastructure(Infrastructure infrastructure) {
		this.infrastructure = infrastructure;
	}

	public Resources resources() {
		return resources;
	}

	public void setResources(Resources resources) {
		this.resources = resources;
	}

//	@Override
//	public int population() {
//		return infrastructure.population();
//	}

	public int aluminum() {
		return resources.get("aluminum").amt();
	}

	public int chromium() {
		return resources.get("chromium").amt();
	}

	public int oil() {
		return resources.get("oil").amt();
	}

	public int rubber() {
		return resources.get("rubber").amt();
	}

	public int steel() {
		return resources.get("steel").amt();
	}

	public int tungsten() {
		return resources.get("tungsten").amt();
	}

	private double tungstenPercentOfGlobal() {
		return (double) tungsten() / State.resourcesOfStates().get("tungsten").amt();       // todo this should really be optimized and some other things later on.
		// todo this all (getting the resources) should be done a lil differently (more generically still.)
	}

	private double steelPercentOfGlobal() {
		// ! states resources and everything are correct, steel() etc must probably be returning 0.
		return (double) steel() / State.resourcesOfStates().get("steel").amt();
	}

	private double rubberPercentOfGlobal() {
		return (double) rubber() / State.resourcesOfStates().get("rubber").amt();
	}

	private double oilPercentOfGlobal() {
		return (double) oil() / State.resourcesOfStates().get("oil").amt();
	}

	private double chromiumPercentOfGlobal() {
		return (double) chromium() / State.resourcesOfStates().get("chromium").amt();
	}

	private double aluminumPercentOfGlobal() {
		return (double) aluminum() / State.resourcesOfStates().get("aluminum").amt();
	}

	public static <T> List<Country> loadCountries(List<T> list) {
		countryList.clear();

		for (T item : list) {
			countryList.add(new Country(item));
		}

		return countryList;
	}

	public static ObservableList<Country> loadCountries(List<CountryTag> countryTags, List<Infrastructure> infrastructureList, List<Resources> resourcesList) {
		countryList.clear();

		Iterator<CountryTag> countryTagsIterator = countryTags.iterator();
		Iterator<Infrastructure> infrastructureListIterator = infrastructureList.iterator();
		Iterator<Resources> resourcesListIterator = resourcesList.iterator();

		while (countryTagsIterator.hasNext()) {
			countryList.add(new Country(countryTagsIterator.next(), infrastructureListIterator.next(), resourcesListIterator.next()));
		}

		return countryList;
	}

	public static ObservableList<Country> loadCountries(List<Infrastructure> infrastructureList, List<Resources> resourcesList) {
		// ! todo resourcesList.get(0).get([resource]).amt() is 0.
		return loadCountries(CountryTags.getCountryTags(), infrastructureList, resourcesList);
	}

	public static ObservableList<Country> loadCountries() {
		return loadCountries(State.infrastructureOfCountries(), State.resourcesOfCountries());
	}

	public String name() {
		return countryTag.toString();
	}

	private int numOwnedStates() {
		return 1;   // todo;
	}

	@Override
	public int compareTo(@NotNull Country o) {
		return countryTag.compareTo(o.countryTag);
	}
}

File: src/main/scala/com/hoi4utils/clausewitz/data/country/Country.scala
package com.hoi4utils.clausewitz.data.country

import com.hoi4utils.clausewitz.data.technology.Technology
import com.hoi4utils.clausewitz.data.units.OrdersOfBattle
import com.hoi4utils.clausewitz.map.buildings.Infrastructure
import com.hoi4utils.clausewitz.map.state.{InfrastructureData, Resource, State}
import com.hoi4utils.clausewitz.script.ReferencePDX
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import org.jetbrains.annotations.NotNull

import scala.jdk.javaapi.CollectionConverters
import java.util
import java.util.function.Function
import scala.collection.mutable.ListBuffer


// todo make country extend countrytag???? ehhhhh
// todo consider... implements infrastructure, resources?????
// todo localizable data?
object Country {
  private val countries = new ListBuffer[Country]()

  def list: List[Country] = countries.toList
  
  def read(): Boolean = {
    // todo
    false
  }

  /**
   * Clears all countries and any other relevant values. 
   */
  def clear(): Unit = {
    countries.clear()
  }

  /**
   * Adds a focus tree to the list of focus trees.
   *
   * @param focusTree the focus tree to add
   * @return the updated list of focus trees
   */
  def add(country: Country): Iterable[Country] = {
    countries += country
    countries
  }

  def getDataFunctions(resourcePercentages: Boolean): Iterable[Country => ?] = {
    val dataFunctions = ListBuffer[Country => ?]()

    dataFunctions += (c => c.name)
    dataFunctions += (c => c._infrastructure.population)
    dataFunctions += (c => c._infrastructure.civilianFactories)
    dataFunctions += (c => c._infrastructure.militaryFactories)
    dataFunctions += (c => c._infrastructure.navalDockyards)
    dataFunctions += (c => c._infrastructure.airfields)
    dataFunctions += (c => c._infrastructure.civMilRatio)
    dataFunctions += (c => c._infrastructure.popPerFactoryRatio)
    dataFunctions += (c => c._infrastructure.popPerCivRatio)
    dataFunctions += (c => c._infrastructure.popPerMilRatio)
    dataFunctions += (c => c._infrastructure.popAirportCapacityRatio)
    dataFunctions += (c => c._infrastructure.popPerStateRatio(c.numOwnedStates))
    // again, this resource code is not expandable. fix sometime :(
    if (resourcePercentages) {
      dataFunctions += (c => c.aluminum)
      dataFunctions += (c => c.chromium)
      dataFunctions += (c => c.oil)
      dataFunctions += (c => c.rubber)
      dataFunctions += (c => c.steel)
      dataFunctions += (c => c.tungsten)
    }
    else {
      dataFunctions += (c => c.aluminumPercentOfGlobal)
      dataFunctions += (c => c.chromiumPercentOfGlobal)
      dataFunctions += (c => c.oilPercentOfGlobal)
      dataFunctions += (c => c.rubberPercentOfGlobal)
      dataFunctions += (c => c.steelPercentOfGlobal)
      dataFunctions += (c => c.tungstenPercentOfGlobal)
    }
    dataFunctions
  }

//  def loadCountries[T](list: List[T]): List[Country] = {
//    countryList.clear()
//
//    for (item <- list) {
//      countryList.add(new Country(item))
//    }
//
//    countryList
//  }

//  def loadCountries(countryTags: List[CountryTag], infrastructureList: List[Infrastructure],
//                    resourcesList: List[List[Resource]]): ObservableList[Country] = {
//    countries.clear()
//
//    val countryTagsIterator = countryTags.iterator
//    val infrastructureListIterator = infrastructureList.iterator
//    val resourcesListIterator = resourcesList.iterator
//
//    while (countryTagsIterator.hasNext) countries.add(new Country(countryTagsIterator.next, infrastructureListIterator.next, resourcesListIterator.next))
//    countries
//  }
//
//  def loadCountries(infrastructureList: List[Infrastructure], resourcesList: List[List[Resource]]): ObservableList[Country] = {
//    loadCountries(CountryTagsManager.getCountryTags, infrastructureList, resourcesList)
//  }

//  def loadCountries: ObservableList[Country] = loadCountries(CollectionConverters.asJava(State.infrastructureOfCountries), CollectionConverters.asJava(State.resourcesOfCountries))
}

class Country extends InfrastructureData with Comparable[Country] {
  private var _countryTag: Option[CountryTag] = None
  private var _infrastructure: Infrastructure = null // infrastructure of all owned states

  private var _resources: List[Resource] = List.empty // resources of all owned states

  private val oob_list: Set[OrdersOfBattle] = null // set of potential orders of battles defined in history/countries file (oob)

  private val defaultResearchSlots = 0 // default research slots as defined in history/countries file or similar

  private val countryFlags: Set[CountryFlag] = null
  private val capital = 0 // country capital as defined by applicable oob or unknown

  private val stability = .0 // stability percentage defined from 0.0-1.0

  private val warSupport = .0 // war support percentage defined from 0.0-1.0

  private val startingTech: Set[Technology] = null // starting technology defined in history/countries file

  def this(countryTag: CountryTag, infrastructure: Infrastructure, resources: List[Resource]) = {
    this()
    this._countryTag = Some(countryTag)
    this._infrastructure = infrastructure
    this._resources = resources
  }

  def this(countryTag: CountryTag) = {
    this(countryTag, new Infrastructure, List[Resource]())
  }

  def countryTag: CountryTag = _countryTag.getOrElse(CountryTag.NULL_TAG)

  def setCountryTag(countryTag: CountryTag): Unit = {
    this._countryTag = Some(countryTag)
  }

  override def getInfrastructureRecord: Infrastructure = _infrastructure

  def setInfrastructure(infrastructure: Infrastructure): Unit = {
    this._infrastructure = infrastructure
  }

  def resources: List[Resource] = _resources

  def setResources(resources: Nothing): Unit = {
    this._resources = resources
  }

  def aluminum: Double = _resources.filter(_.isValidID("aluminum")).map(_.amt).sum

  def chromium: Double = _resources.filter(_.isValidID("chromium")).map(_.amt).sum

  def oil: Double = _resources.filter(_.isValidID("oil")).map(_.amt).sum

  def rubber: Double = _resources.filter(_.isValidID("rubber")).map(_.amt).sum

  def steel: Double = _resources.filter(_.isValidID("steel")).map(_.amt).sum

  def tungsten: Double = _resources.filter(_.isValidID("tungsten")).map(_.amt).sum

  private def tungstenPercentOfGlobal = {
    tungsten / State.resourcesOfStates.filter(_.isValidID("tungsten")).map(_.amt).sum

    // todo this all (getting the resources) should be done a lil differently (more generically still.)
  }

  private def steelPercentOfGlobal = {
    // ! states resources and everything are correct, steel() etc must probably be returning 0.
    tungsten / State.resourcesOfStates.filter(_.isValidID("steel")).map(_.amt).sum
  }

  private def rubberPercentOfGlobal = tungsten / State.resourcesOfStates.filter(_.isValidID("rubber")).map(_.amt).sum

  private def oilPercentOfGlobal = tungsten / State.resourcesOfStates.filter(_.isValidID("oil")).map(_.amt).sum

  private def chromiumPercentOfGlobal = tungsten / State.resourcesOfStates.filter(_.isValidID("chromium")).map(_.amt).sum

  private def aluminumPercentOfGlobal = tungsten / State.resourcesOfStates.filter(_.isValidID("aluminum")).map(_.amt).sum

  def name: String = _countryTag.toString

  private def numOwnedStates = 1 // todo;

  override def compareTo(@NotNull o: Country): Int = _countryTag match {
    case Some(tag) => tag.compareTo(o.countryTag)
    case None => 0
  }

  override def toString: String = _countryTag.toString + " " + "[country tag]"
}
